xds/src/main/java/io/grpc/xds/EnvoyProtoData.java

```diff
@@ -339,4 +340,194 @@ public String toString() {
           .toString();
     }
   }
+
+  /** See corresponding Envoy proto message {@link io.envoyproxy.envoy.api.v2.route.Route}. */
+  static final class Route {
+    final RouteMatch routeMatch;
+    final Optional<RouteAction> routeAction;
```

Comment:
The benefit: 
https://github.com/google/guava/wiki/UsingAndAvoidingNullExplained#whats-the-point

That might be opinionated as well as the SO.

Reply:
According to its [Javadoc](https://guava.dev/releases/19.0/api/docs/com/google/common/base/Optional.html), the most common usages for `Optional` are:
- __As a method return type__, as an alternative to returning null to indicate that no value was available
- To __distinguish between "unknown" (for example, not present in a map) and "known to have no value" (present in the map, with value Optional.absent())__
- To __wrap nullable references for storage in a collection that does not support null__ (though there are several other approaches to this that should be considered first)

It's not for usage of avoiding nullness just for convenience. This usage is completely an abuse, even your IDE will warn you "Optional used as type for field". For details, read the first item of [Optional Anti-Patterns](https://dzone.com/articles/optional-anti-patterns).

Reply:
For the IDE waring and Optional Anti-Patterns, it's only a problem when you want to serialize the class. The usage here is the first one in the javadoc, though the field can be just nullable to be exactly the case.

Reply:
"Anti-pattern" indicates it's a rule-of-thumb convention that you should avoid, even though it may not cause problem now. It's a bad practice to do that and it's more likely to cause problem in the future. I do not see any benefit than a nullable field (which is more elegant and efficient). Why using `Optional` as a class field type is the first category of usage (i.e., as return type)? The first category of usage is referring to the following pattern:

```java
class Foo {
   // nullable
   private Bar bar;

  Optional<Bar> getBar() {
     // ...
  }
}
```